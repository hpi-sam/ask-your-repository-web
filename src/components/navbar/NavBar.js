// @flow
import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';
import { MdCloudUpload, MdImage } from 'react-icons/md';
import { NavLink } from 'react-router-dom';
import Search from './Search';
import Dropdown from './Dropdown';
import TeamInitialsButton from '../team/TeamInitialsButton';
import PresentationSwitch from './PresentationSwitch';
import type { Team } from '../../models/Team';
import type { AppState } from '../../state/AppState';
import { openTeamSidebar } from '../../state/team_sidebar/teamSidebar.actionCreators';
import './NavBar.scss';

type Props = {
  activeTeam: ?Team,
  isTeamSidebarOpen: boolean,
  isAuthenticated: boolean,
  onTeamClick: () => void,
};

function NavBar(props: Props) {
  const {
    isTeamSidebarOpen, onTeamClick, activeTeam, isAuthenticated,
  } = props;

  return (
    <div className={classNames('NavBar', { 'NavBar--with-sidebar': isTeamSidebarOpen })}>
      <div className="NavBar__left">
        {!isTeamSidebarOpen && activeTeam && (
          <TeamInitialsButton
            onClick={onTeamClick}
            className="NavBar__team"
            team={activeTeam}
            data-cy="navbar-team-button"
          />
        )}
        <NavLink
          to="/images"
          className="NavBar__item"
          activeClassName="NavBar__item--active"
        >
          <MdImage className="NavBar__item__icon" />
          <span className="NavBar__item__text">
            Gallery
          </span>
        </NavLink>
        <NavLink
          to="/upload"
          className="NavBar__item"
          activeClassName="NavBar__item--active"
        >
          <MdCloudUpload className="NavBar__item__icon" />
          <span className="NavBar__item__text">
            Upload
          </span>
        </NavLink>
        <PresentationSwitch />
      </div>
      <div className="NavBar__search">
        <Search />
      </div>
      <div className="NavBar__right">
        {isAuthenticated
          ? <Dropdown />
          : (
            <Fragment>
              <NavLink
                to="/login"
                className="NavBar__item"
                activeClassName="NavBar__item--active"
              >
                Login
              </NavLink>
              <NavLink
                to="/register"
                className="NavBar__item"
                activeClassName="NavBar__item--active"
              >
                Sign up
              </NavLink>
            </Fragment>
          )
        }
      </div>
    </div>
  );
}

const mapDispatchToProps = (dispatch: Function) => ({
  onTeamClick: () => { dispatch(openTeamSidebar()); },
});

const mapStateToProps = (state: AppState) => ({
  activeTeam: state.activeTeam,
  isTeamSidebarOpen: state.teamSidebar.isOpen,
  isAuthenticated: state.auth.loggedIn,
});

export default connect(mapStateToProps, mapDispatchToProps)(NavBar);
