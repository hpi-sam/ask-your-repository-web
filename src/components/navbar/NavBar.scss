@import '../../style/breakpoints.scss';

:root {
  --navbar-height: 64px;
  --navbar-padding: 10px;
}

.NavBar {
  top: 0;
  left: 0;
  right: 0;
  position: fixed;
  z-index: var(--navbar-index);

  display: flex;
  flex-direction: column;
  background-color: var(--navbar-bg-color);

  @media (min-width: $sm-to-md) {
    transition: left var(--sidebar-animation-duration);
    &--with-sidebar { left: var(--team-sidebar-width); }
  }

  &__inner {
    display: flex;
    align-items: center;
    height: var(--navbar-height);
    padding: var(--navbar-padding);
  }

  &__left, &__right {
    display: flex;
    height: 100%;
  }

  &__left {
    flex: 1;
  }

  &__left &__item {
    padding: 0 30px;
  }

  &__search {
    flex: 1;
  }

  &__presentation-switch {
    display: none;

    @media (min-width: $sm-to-md) {
      display: flex;
    }
  }

  &__right &__item {
    margin-left: 0;
    margin-right: 0;
    padding-left: 30px;

    &:last-child {
      margin-right: 20px;
    }
  }

  &__team {
    margin-right: 10px;
    align-self: center;
  }

  &__item {
    display: flex;
    align-items: center;
    font-weight: bold;
    color: var(--white);

    &:hover {
      .navbar-key {
        color: var(--white-hover)
      }
    }

    &--active {
      color: var(--primary-color);

      &:hover {
        .navbar-key {
          color: var(--primary-color-hover);
        }
      }
    }

    &__text {
      display: none;
      margin-left: 10px;

      @media (min-width: $xs-to-sm) {
        display: inline;
      }
    }

    &__icon {
      font-size: 1.5em;
    }
  }
}
