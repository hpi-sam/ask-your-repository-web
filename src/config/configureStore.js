// @flow
import { applyMiddleware, createStore } from 'redux';
import type { Dispatch } from 'redux';
import { persistStore, persistReducer } from 'redux-persist-erksch';
import storage from 'redux-persist-erksch/lib/storage';
import { routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { middleware as flashMiddleware } from 'redux-flash';
import thunk from 'redux-thunk';
import socketioMiddleware from '../middleware/socketioMiddleware';
import axiosAuthMiddleware from './axiosAuthMiddleware';
import createRootReducer from '../state/rootReducer';
import type { AppState } from '../state/AppState';
import type { Action } from '../state/Action';

export const history = createBrowserHistory();

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['activeTeam', 'presentationMode', 'auth'],
};

function configureStore() {
  let middleware = [
    thunk,
    flashMiddleware(),
    routerMiddleware(history),
    socketioMiddleware(),
    axiosAuthMiddleware(),
  ];

  if (process.env.NODE_ENV === 'development') {
    const { logger } = require('redux-logger');
    middleware = [...middleware, logger];
  }

  const rootReducer = createRootReducer(history);
  const persistedReducer: any = persistReducer(persistConfig, rootReducer);

  const store = createStore<AppState, Action, Dispatch<Action>>(
    persistedReducer,
    applyMiddleware(...middleware),
  );

  const persistor = persistStore(store);

  return { store, persistor };
}

export default configureStore;
