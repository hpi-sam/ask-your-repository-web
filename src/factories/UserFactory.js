// @flow
import faker from 'faker';
import type { User } from '../models/User';

class UserFactory {
  static createDummyUser(): User {
    return {
      id: faker.random.uuid(),
      email: faker.internet.email(),
      username: faker.internet.userName(),
    };
  }

  static createAuthenticatedUser(): User {
    return {
      ...this.createDummyUser(),
      token: faker.random.uuid(),
    };
  }

  static createStaticDummyUser(): User {
    return {
      id: 'f1bc8f40-7698-4df7-b410-efac1f1a03d2',
      email: 'test@example.com',
      username: 'test_user',
    };
  }

  static createStaticAuthenticatedUser(): User {
    return {
      ...this.createStaticDummyUser(),
      token: '61e6fa4c-de31-48e6-97a2-4f66b31d84e9',
    };
  }
}

export default UserFactory;
