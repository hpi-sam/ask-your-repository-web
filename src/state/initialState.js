// @flow

// This file is only for state related testing

import type { AppState } from './AppState';
import { initialState as presentation } from './presentation/presentation.reducer';
import { initialState as teamSidebar } from './team_sidebar/teamSidebar.reducer';
import { initialState as presentationMode } from './presentation_mode/presentationMode.reducer';
import { initialState as activeTeam } from './active_team/activeTeam.reducer';
import { initialState as auth } from './auth/auth.reducer';

const initialState: AppState = {
  flash: { messages: [] },
  presentation,
  activeTeam,
  teamSidebar,
  presentationMode,
  auth,
};

export default initialState;
