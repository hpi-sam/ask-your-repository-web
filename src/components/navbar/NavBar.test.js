// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import initialState from '../../state/initialState';
import UserFactory from '../../factories/UserFactory';
import TeamFactory from '../../factories/TeamFactory';
import NavBar from './NavBar';

const mockStore = configureStore();

const state = {
  ...initialState,
  user: UserFactory.createStaticAuthenticatedUser(),
  activeTeam: TeamFactory.createStaticDummyTeam(),
};

describe('<NavBar />', () => {
  let wrapper;
  let store;

  beforeEach(() => {
    store = mockStore(state);
    wrapper = mount((
      <Provider store={store}>
        <MemoryRouter>
          <NavBar />
        </MemoryRouter>
      </Provider>
    )).find('NavBar');
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
