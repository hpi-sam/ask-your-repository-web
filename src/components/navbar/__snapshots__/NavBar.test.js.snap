// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar /> renders correctly 1`] = `
<NavBar
  activeTeam={null}
  isTeamSidebarOpen={false}
  onTeamClick={[Function]}
>
  <div
    className="NavBar"
  >
    <div
      className="NavBar__inner"
    >
      <div
        className="NavBar__right"
      >
        <NavLink
          activeClassName="NavBar__item--active"
          aria-current="page"
          className="NavBar__item"
          to="/images"
        >
          <Route
            path="\\\\/images"
          >
            <Link
              aria-current={null}
              className="NavBar__item"
              replace={false}
              to="/images"
            >
              <a
                aria-current={null}
                className="NavBar__item"
                href="/images"
                onClick={[Function]}
              >
                <MdImage
                  className="NavBar__item__icon navbar-key"
                >
                  <IconBase
                    attr={
                      Object {
                        "viewBox": "0 0 24 24",
                      }
                    }
                    className="NavBar__item__icon navbar-key"
                  >
                    <svg
                      className="NavBar__item__icon navbar-key"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
                        key="0"
                      />
                    </svg>
                  </IconBase>
                </MdImage>
                <span
                  className="NavBar__item__text navbar-key"
                >
                  Gallery
                </span>
              </a>
            </Link>
          </Route>
        </NavLink>
        <div
          className="NavBar__item"
        >
          <Connect(OnClickOutside(Dropdown))>
            <OnClickOutside(Dropdown)
              dispatch={[Function]}
              eventTypes={
                Array [
                  "mousedown",
                  "touchstart",
                ]
              }
              excludeScrollbar={false}
              outsideClickIgnoreClass="ignore-react-onclickoutside"
              preventDefault={false}
              stopPropagation={false}
              user={null}
            >
              <Dropdown
                disableOnClickOutside={[Function]}
                dispatch={[Function]}
                enableOnClickOutside={[Function]}
                eventTypes={
                  Array [
                    "mousedown",
                    "touchstart",
                  ]
                }
                outsideClickIgnoreClass="ignore-react-onclickoutside"
                preventDefault={false}
                stopPropagation={false}
                user={null}
              >
                <div
                  className="Dropdown"
                >
                  <div
                    className="Dropdown__inner"
                  >
                    <button
                      className="Dropdown__button"
                      data-cy="user-dropdown-button"
                      onClick={[Function]}
                      type="button"
                    >
                      <MdPerson
                        className="Dropdown__button__icon navbar-key"
                      >
                        <IconBase
                          attr={
                            Object {
                              "viewBox": "0 0 24 24",
                            }
                          }
                          className="Dropdown__button__icon navbar-key"
                        >
                          <svg
                            className="Dropdown__button__icon navbar-key"
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            strokeWidth="0"
                            style={
                              Object {
                                "color": undefined,
                              }
                            }
                            viewBox="0 0 24 24"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                              key="0"
                            />
                          </svg>
                        </IconBase>
                      </MdPerson>
                      <span
                        className="Dropdown__button__username navbar-key"
                      />
                      <IoIosArrowDown
                        className="navbar-key"
                      >
                        <IconBase
                          attr={
                            Object {
                              "viewBox": "0 0 512 512",
                            }
                          }
                          className="navbar-key"
                        >
                          <svg
                            className="navbar-key"
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            strokeWidth="0"
                            style={
                              Object {
                                "color": undefined,
                              }
                            }
                            viewBox="0 0 512 512"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                              key="0"
                            />
                          </svg>
                        </IconBase>
                      </IoIosArrowDown>
                    </button>
                    <div
                      className="Dropdown__container"
                    >
                      <div
                        className="Dropdown__arrow"
                      />
                      <div
                        className="Dropdown__content"
                      >
                        <ButtonLink
                          className="Dropdown__content__button"
                          to="/settings"
                        >
                          <Link
                            className="Button Dropdown__content__button"
                            replace={false}
                            to="/settings"
                          >
                            <a
                              className="Button Dropdown__content__button"
                              href="/settings"
                              onClick={[Function]}
                            >
                              Settings
                            </a>
                          </Link>
                        </ButtonLink>
                        <Button
                          className="Dropdown__content__button"
                          onClick={[Function]}
                          type="button"
                        >
                          <button
                            className="Button Dropdown__content__button"
                            onClick={[Function]}
                            type="button"
                          >
                            Logout
                          </button>
                        </Button>
                      </div>
                    </div>
                  </div>
                </div>
              </Dropdown>
            </OnClickOutside(Dropdown)>
          </Connect(OnClickOutside(Dropdown))>
        </div>
        <div
          className="NavBar__item NavBar__presentation-switch"
        >
          <Connect(PresentationSwitch)>
            <PresentationSwitch
              dispatch={[Function]}
              isPresentationModeOn={false}
            >
              <button
                className="PresentationSwitch"
                data-tip="Synchronized Presentation Mode"
                onClick={[Function]}
                type="button"
              >
                <div
                  className="PresentationSwitch__outer"
                >
                  <div
                    className="PresentationSwitch__inner"
                  />
                </div>
              </button>
              <ReactTooltip
                clickable={false}
                effect="solid"
                insecure={true}
                place="bottom"
                resizeHide={true}
                wrapper="div"
              >
                <div
                  className="__react_component_tooltip place-bottom type-dark "
                  data-id="tooltip"
                />
              </ReactTooltip>
            </PresentationSwitch>
          </Connect(PresentationSwitch)>
        </div>
      </div>
    </div>
  </div>
</NavBar>
`;
